### PR TITLE
Improve memory usage on software channels export

### PR DESCRIPTION
**Pull request by rjmateus**
Followup of: https://github.com/SUSE/spacewalk/issues/16963
Implementation of: https://github.com/SUSE/spacewalk/issues/17193
Improve memory usage in datacrawler.

This can be done by reducing the type of structure used in [1] which is responsible for 80%+ of the memory usage.

image

## Results
- Channel label: sle-module-basesystem15-updates-x86_64-sled
- Time in DataCrawler: 
  - Base case: ~10min(+-30sec), from 14:34:52 to 14:44:52
  - Improved case: ~10min(+-30sec), from 14:54:44 to 15:04:44
- Peak memory usage:
  - Base case ~1433mb
  - Improved case ~543mb

Before:
image

After:
image

[1] https://github.com/uyuni-project/inter-server-sync/blob/main/dumper/types.go#L9-L13

Signed-off-by: Ricardo Mateus <rmateus@suse.com>

### CODE REVIEW

**Comment by avshiliaev** [comment]
@rjmateus Added some more changes and rebased. Please take a look :smiley:

**Comment by rjmateus** [comment]
@aaannz @avshiliaev this PR is ready for review. I have tested it locally several times and looks to be working as expected.
I also asked for feedback from Anne-Marie who will run a small test more from a usability perspective.

File: entityDumper/channelDataDumper.go [hunk 1]
```diff
@@ -201,7 +203,13 @@ func processChannel(db *sql.DB, writer *bufio.Writer, channelLabel string,
...
 	whereFilter := fmt.Sprintf("label = '%s'", channelLabel)
 	tableData := dumper.DataCrawler(db, schemaMetadata, schemaMetadata["rhnchannel"], whereFilter, options.StartingDate)
-	log.Debug().Msg("finished table data crawler")
+
+	totalRows := 0
+	for _, value := range tableData.TableData {
+		totalRows = totalRows + len(value.KeyMap)
+	}
+
+	log.Debug().Msgf("finished table data crawler. Total database rows to export: %d", totalRows)
```
Comment (aaannz) [inline entityDumper/channelDataDumper.go:212]: Nitpick, but this should have been under `if log.Debug().Enabled() {` i.e.

```suggestion
	if log.Debug().Enabled() {
		totalRows := 0
		for _, value := range tableData.TableData {
			totalRows = totalRows + len(value.KeyMap)
		}
		log.Debug().Msgf("finished table data crawler. Total database rows to export: %d", totalRows)
	}
```

**Review by aaannz — APPROVED** [review]
LGTM.

IIUC improvement comes from instead of doing recursive calls we now do all in for loop.
Changes in `Table` type are more-so cosmetic, or it there some go optimization going on?

**Comment by rjmateus** [comment]
@aaannz the biggest improvement comes from the change from recursive to iterative processing.
However changing also the data structure from map to array also improves the memory usage, since maps in Go uses much more memory then arrays.

**Review by avshiliaev — APPROVED** [review]
LGTM everything looks sensibel and neat :smiley:

**Pull request merged**